Add CLI for dumping BPF endpoint map (lxcmap)

joestringer opened the pull request:
This series adds a command `cilium bpf endpoint list` which dumps all entries in the `lxcmap`, for debugging purposes. It also does some other misc cleanups in this area.

joestringer committed:
lxcmap: Replace IPv6 address field with padding

This field was unused, replace it with padding and don't attempt to
populate or format its contents.

Signed-off-by: Joe Stringer <joe@covalent.io>

joestringer committed:
lxcmap: Format host values as "(localhost)"

The string-representation of lxcmap entries that purely identify that
the address belongs to the local host is overly verbose, given that it
only holds one bit of information: The fact that the IP belongs to the
host. Return a simple string representation for this case.

Signed-off-by: Joe Stringer <joe@covalent.io>

joestringer committed:
cmd: Add 'bpf endpoint list' command

Previously we had no convenient way to access the bpf "lxcmap" contents
for debugging the datapath. This commit adds support for dumping these
entries.

Signed-off-by: Joe Stringer <joe@covalent.io>

aanm commented on common/types/ipv6.go:
```diff
@@ -25,6 +25,12 @@ func (v6 IPv6) IP() net.IP {
+// To4 converts this address into an IPv4 address, for areas of the code which
+// reuse this structure for both IPv4 and IPv6 addresses.
+func (v6 IPv6) To4() net.IP {
+	return v6[:4]
```
This needs to be documented or changed. Golang can use a 16 byte array for v4 addresses but they store it at the end of the array. https://golang.org/src/net/ip.go#L45 and https://golang.org/src/net/ip.go#L186

Can we do the same?

ianvernon replied:
Seconded, we should maintain consistency with golang.

joestringer replied:
This IPv6 address structure is intended to match the BPF representation, and should perhaps be instead placed somewhere like `pkg/bpf/types.go`. The implication of making such a change also involves updating the BPF code to understand the new format, and potentially impacts upgrade.

joestringer replied:
Looking closer, the referred `IPv4(a, b, c, d)` will actually return a 16-byte array with the IPv4 address in the highest four bytes, but it's not consistent - look at https://golang.org/src/net/ip.go#L59 . Weirdly, if you then perform `String()` on an address created using `IPv4()`, it will be a 16-byte array so will be formatted as an IPv6 address. The `net.IP` type itself is just a `[]byte`, which means that in most cases an IPv4 address should be represented as `[4]byte`, not `[16]byte` with the address in the uppermost four bytes.

aanm commented on pkg/maps/lxcmap/lxcmap.go:
```diff
@@ -141,25 +143,35 @@ func (k EndpointKey) GetKeyPtr() unsafe.Pointer { return unsafe.Pointer(&k) }
+func (k EndpointKey) String() string {
+	switch k.Family {
+	case endpointKeyIPv4:
+		return k.IP.To4().String()
```
Golang has this method (`To4()`) and can return nil, I know this is our method but this goes with my concerns on using the same methods with different internal implementation (the v4 address being stored on a different location on the 16 byte array)

joestringer replied:
I see that for the `lxcmap` and the `tunnel` map, they both actually hold an `IP [16]byte` followed by `Family uint8`. Perhaps I should look into defining a new IP type that handles this better for the BPF maps which use the same memory to represent either IPv4 or IPv6 addresses.

aanm commented on pkg/maps/lxcmap/lxcmap.go:
```diff
@@ -217,3 +234,30 @@ func DeleteElement(f EndpointFrontend) int {
+	}
+
+	if err := binary.Read(bytes.NewBuffer(value), byteorder.Native, &v); err != nil {
+		return nil, nil, fmt.Errorf("Unable to convert key: %s", err)
```
s/` key`/` value`

joestringer replied:
Will fix, thanks for the review.

ianvernon reviewed (dismissed):
Please address the comments. Looks fine otherwise. Thanks for adding this - this will be another helpful datapoint for debugging if we hit issues! Can you add a GitHub issue to add a runtime test that tests this ?

joestringer committed:
bpf: Refactor EndpointKey

Endpoint keys, including tunnel keys, use a common structure like this:

{
	union {
		struct {
			__u32		ip4;
			__u32		pad1;
			__u32		pad2;
			__u32		pad3;
		};
		union v6addr	ip6;
	};
	__u8 family;
	__u8[3] pad;
}

This commit introduces a common type to handle this in the golang code,
with converters into the standard golang net.IP type. Note that the IPv4
address differs in storage format from the standard net.IP - using the
lowest four bytes rather than the highest.

This new type reduces the amount of changes required in an upcoming
commit to format lxcmap entries.

Signed-off-by: Joe Stringer <joe@covalent.io>

joestringer committed:
tunnel: Fix typo

Signed-off-by: Joe Stringer <joe@covalent.io>

tgraf reviewed (dismissed):
Needs rebase
